fl4health/clients/adaptive_drift_constraint_client.py

```diff
@@ -0,0 +1,214 @@
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        checkpointer: Optional[ClientCheckpointModule] = None,
+        metrics_reporter: Optional[MetricsReporter] = None,
+        progress_bar: bool = False,
+    ) -> None:
+        """
+        This client serves as a base for FL methods implementing an auxiliary loss penalty with a weight coefficient
+        that might be adapted via loss trajectories on the server-side. An example of such a method is FedProx, which
+        uses an auxiliary loss penalizing weight drift with a coefficient mu. This client is a simple extension of the
+        BasicClient that packs the criterion loss (i.e. loss without the penalty) for exchange with the server and
+        expects to receive an updated (or constant if non-adaptive) parameter for the loss weight.
+
+        Args:
+            data_path (Path): path to the data to be used to load the data for client-side training
+            metrics (Sequence[Metric]): Metrics to be computed based on the labels and predictions of the client model
+            device (torch.device): Device indicator for where to send the model, batches, labels etc. Often 'cpu' or
+                'cuda'
+            loss_meter_type (LossMeterType, optional): Type of meter used to track and compute the losses over
+                each batch. Defaults to LossMeterType.AVERAGE.
+            checkpointer (Optional[ClientCheckpointModule], optional): Checkpointer module defining when and how to
+                do checkpointing during client-side training. No checkpointing is done if not provided. Defaults to
+                None.
+            metrics_reporter (Optional[MetricsReporter], optional): A metrics reporter instance to record the metrics
+                during the execution. Defaults to an instance of MetricsReporter with default parameters.
+            progress_bar (bool): Whether or not to display a progress bar during client training and validation.
+                Uses tqdm. Defaults to False
+        """
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            checkpointer=checkpointer,
+            metrics_reporter=metrics_reporter,
+            progress_bar=progress_bar,
+        )
+        # These are the tensors that will be used to compute the penalty loss
+        self.drift_penalty_tensors: List[torch.Tensor]
+        # Exchanger with packing to be able to exchange the weights and auxiliary information with the server for
+        # adaptation
+        self.parameter_exchanger: FullParameterExchangerWithPacking
+        # Weight on the penalty loss to be used in backprop. This is what might be adapted via server calculations
+        self.drift_penalty_weight: float
+        # This is the loss value to be sent back to the server on which adaptation decisions will be made.
+        self.loss_for_adaptation: float
```

Comment:
nit: In the docstring above, you refer to this as: "the criterion loss (i.e. loss without the penalty)". Maybe its worth updating the language to be more consistent so its clear that is the loss that is being referred to

Reply:
Added some language in the init docstring to better tie these together. It doesn't necessarily have to be the criterion loss. So I made the docstring for the class more general.